src/renderer/components/dock/terminal/store.ts

```diff
@@ -20,69 +20,71 @@ interface Dependencies {
   createTerminalApi: CreateTerminalApi;
 }
 
+export interface TerminalConnection {
+  terminal: Terminal;
+  api: TerminalApi;
+}
+
+export interface TerminalConnectOptions {
+  signal: AbortSignal;
+}
+
 export class TerminalStore {
-  protected terminals = new Map<TabId, Terminal>();
-  protected connections = observable.map<TabId, TerminalApi>();
+  private readonly connections = new Map<TabId, TerminalConnection>();
 
-  constructor(private dependencies: Dependencies) {
-  }
+  constructor(private readonly dependencies: Dependencies) {}
 
-  @action
-  connect(tab: ITerminalTab) {
-    if (this.isConnected(tab.id)) {
-      return;
+  connect(tab: ITerminalTab): TerminalConnection & { connectionPromise?: Promise<void> } {
+    {
+      const connection = this.connections.get(tab.id);
+
+      if (connection) {
+        return connection;
+      }
     }
+
     const api = this.dependencies.createTerminalApi({
       id: tab.id,
       node: tab.node,
     });
     const terminal = this.dependencies.createTerminal(tab.id, api);
 
-    this.connections.set(tab.id, api);
-    this.terminals.set(tab.id, terminal);
+    runInAction(() => {
+      this.connections.set(tab.id, { api, terminal });
+    });
+
+    const connectionPromise = api.connect();
```

Comment:
Why are we not `await`ing for promise?

Reply:
So that the users of `connect` can be sync for when they want to do things that are not related to the connection being fully active.